Added a text input that only accepts full numbers (int)

leedave opened the pull request:
# What's new

- I've added a new Keyboard input. It is similar to text_input and byte_input. The purpose of this one is to only allow users to input a number, nothing else. This can be very handy in app designs, as the user input is limited to expected characters.

# Verification 

- #include <gui/modules/int_input.h> in a fap application the same way you would include text_input.h

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

Commit:
Added a text input that only accepts full numbers (int)

Commit:
Added a text input that only accepts full numbers (int)

hedger reviewed (dismissed):
Please also include the new header in "sdk_headers" for gui module. And update API CSV's for both f7 and f18 afterwards.

Commit:
Added to Gui sdk_headers and api_symbols in f7 and f18

Commit:
Added to Gui sdk_headers and api_symbols in f7 and f18

leedave commented:
@hedger Thank you for the feedback. 
Added the requested sdk_headers and CSV entries

leedave commented:
@hedger are these fail messages normal or am I expected to do something to fix this?

Commit:
Fixed _Bool declarations in symbols csv

skotopes commented:
Sorry for delay with your PR. 
Can you also provide example app and rename `IntInput` to `NumberInput`, `int_input` to `number_input`?

Commit:
renamed int_input to number_input

Commit:
Changed name & added example fap

Commit:
Changed name & added example fap

leedave commented:
@skotopes 
Ok, tried my best to keep it as expected. Let me know if something isn't right. 

- The keyboard is renamed
- There is an example app in **applications/examples/example_number_input**

leedave commented:
https://github.com/flipperdevices/flipperzero-firmware/assets/5908330/7530c74e-8b17-4775-b99e-4c433c50c6b0

Commit:
formatting code

skotopes commented:
```[Switching to Thread 536956528]
0x0801147c in __furi_crash_implementation () at furi/core/check.c:163
163	        RESTORE_REGISTERS_AND_HALT_MCU(debug);
(gdb) bt
#0  0x0801147c in __furi_crash_implementation () at furi/core/check.c:163
#1  0x0807a5cc in view_dispatcher_free (view_dispatcher=0x20014de8) at applications/services/gui/view_dispatcher.c:26
#2  0x200154a6 in example_number_input_free (app=app@entry=0x20015a00) at applications/examples/example_number_input/example_number_input.c:45
#3  0x2001551e in example_number_input (p=<optimized out>) at applications/examples/example_number_input/example_number_input.c:69
#4  0x080548a2 in flipper_application_thread (context=0x20014b20) at lib/flipper_application/flipper_application.c:224
#5  0x08013d6e in furi_thread_body (context=0x20014d78) at furi/core/thread.c:91
#6  0x08013d1a in furi_thread_catch () at furi/core/thread.c:62
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```

crashing on exit because some view was not released from view dispatcher

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,389 @@
+            model->selected_row = 0;
+            model->selected_column = 0;
+            model->clear_default_text = false;
+            model->text_buffer = "";
```
I got feeling that this part was copied from another place. We are not proud with it )

General concept for view modules is that they should own their own buffer and then provide it in callback call.

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,389 @@
+static const uint8_t enter_symbol = '\r';
+static const uint8_t backspace_symbol = '\b';
+
+static const NumberInputKey keyboard_keys_row_1[] = {
```
Since we talking about numbers(int) you may want to provide negative sign too.
Also it will be nice if limits(MIN/MAX) and maybe int size be adjustable

skotopes commented:
Un-draft when ready

skotopes committed:
Minor clenaup

skotopes committed:
Minor clenaup

skotopes commented:
also please fix build issues (API mismatch)

skotopes commented:
@leedave need help?

WillyJL committed:
GUI: Fix array out of bounds in menu exit (#3604)

* GUI: Fix array out of bounds in menu exit
* Gui: fix incorrect empty menu handling
* Gui: add missing item check in menu ok handling
* Gui: remove dead code from menu module

Co-authored-by: あく <alleteam@gmail.com>

gornekich committed:
nfc app: add legacy keys for plantain cards (#3602)

Co-authored-by: あく <alleteam@gmail.com>

leedave commented:
@skotopes 
Can you give me some more hints on the 
`view modules should own their own buffer` issue. 
Tbh I learn how code for Flipper should look by studying the fw you provide, as there is no official developer documentation. I got the basic setup from the other two keyboards (text & hex). If you say you're not proud of what you guys did back then and ask for an update, I'm not sure where I can find something built in the way you want to see it.

skotopes commented:
@leedave change `model->text_buffer` type to `FuriString` and allocate it on your module allocation

Commit:
update for clearing views

leedave commented:
@skotopes 
How did you do the check, to see if some views were not removed from view_dispatcher. I added a code update, but lack a way of testing the result

Commit:
rebase dev

Commit:
Resolve conflict

leedave committed:
Merge branch 'dev' into leedave/int_text_input

Commit:
refactoring test app, part 1

Commit:
Refactor test app, part 2

leedave committed:
Minor updates while travelling

Commit:
Switched from const char to FuriString. Using Temp module copy for development to spare compile time

Commit:
Option to limit number output with min and max values

Commit:
Preparations for option to change number sign from + to -

Commit:
Preparations for option to change number sign from + to -

leedave committed:
Merge branch 'dev' into leedave/int_text_input

Commit:
Preparing for testing

leedave committed:
Merge branch 'dev' into leedave/int_text_input

Commit:
counter automatic API version change

Commit:
fbt format

leedave commented:
So after a long wait, I finally managed to add your requests

- `model->text_buffer` changed to FuriString
- removed number digits length option
- replaced with settings for min/max value allowed
- added possibility to input negative numbers, added button if set to true when loading module

Commit:
added trailing comma in application.fam ... because the lint check wants it¿

skotopes committed:
Merge branch 'dev' into leedave/int_text_input

skotopes committed:
Merge branch 'dev' into leedave/int_text_input

skotopes commented on targets/f7/api_symbols.csv:
```diff
@@ -20,6 +20,7 @@ Header,+,applications/services/gui/modules/file_browser.h,,
 Header,+,applications/services/gui/modules/file_browser_worker.h,,
 Header,+,applications/services/gui/modules/loading.h,,
```
Bump minor api version

leedave replied:
Not sure if I'm doing this correctly... I assume I need to run `.\fbt` , fix entries with ? and then commit?

skotopes replied:
Change `Version,+,68.1,,` to `Version,+,68.2,,`

skotopes commented on targets/f18/api_symbols.csv:
```diff
@@ -13,6 +13,7 @@ Header,+,applications/services/gui/icon_i.h,,
 Header,+,applications/services/gui/modules/button_menu.h,,
 Header,+,applications/services/gui/modules/button_panel.h,,
```
Bump minor api version

skotopes commented on applications/services/gui/modules/number_input.h:
```diff
@@ -0,0 +1,75 @@
+    NumberInput* number_input,
+    NumberInputCallback input_callback,
+    void* callback_context,
+    FuriString* text_buffer,
```
There is no need for this module to have `FuriString* text_buffer`.

skotopes commented on applications/services/gui/modules/number_input.h:
```diff
@@ -0,0 +1,75 @@
+typedef struct NumberInput NumberInput;
+
+/** callback that is executed on save button press */
+typedef void (*NumberInputCallback)(void* context);
```
Pass number to callback

skotopes commented on applications/services/gui/modules/number_input.h:
```diff
@@ -0,0 +1,75 @@
+typedef void (*NumberInputCallback)(void* context);
+
+/** callback that is executed when byte buffer is changed */
+typedef void (*NumberChangedCallback)(void* context);
```
Pass number or text to callback.
Also what's meaning of this callback? does it works as validator? In that case it should return something to signal validation results.

leedave replied:
You're right, this is not needed. Will be removed

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,493 @@
+
+typedef struct {
+    const char* header;
+    FuriString* text_buffer;
```
Do you need text? At the end you are working with the numbers

leedave replied:
Yes, the input is a keyboard and therefore it creates a string, which is also shown on screen. 
When I follow your change requests to return a number in callback, I'd convert this string and remove it when the module is closed.

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,493 @@
+    int32_t max_value;
+    int32_t min_value;
+    bool useSign;
+    char sign;
```
Too many redundant things do you really need them?

leedave replied:
Please explain, for me these are not redundant, they are all needed. 
max_value => highest number allowed
min_value => lowest number allowed
sign => + or - depending if the number is positive or negative
useSign => if true, will show the button that allows changing from positive to negative numbers 
(In my use cases so far I did not want negative numbers and therefore don't want that button, this was a change request from your side)

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,493 @@
+    char sign;
+    NumberInputCallback callback;
+    void* callback_context;
+    int8_t selected_row;
```
Can it be negative? if not then size_t. In general there is no sense in using uintX_t types where specialized one exists

leedave replied:
Your right, as it cannot be negative, there's no need for a sign. 

I'm having trouble figuring out why **size_t** is preferred over **uint8_t**. Can you explain this? From what I can gather it's just uint8_t is alway 8-Bit while the Bit-size of size_t are platform dependent.

skotopes replied:
It's all about machine word size and architecture. If you have 32bit Arm MCU then it will operate with 32bit wide registers. So basically you place your 8bit value to 32bit register do things and then transfer it back. That's not going to be faster then doing same thing with 32bit wide variable.

Also 8 bit is kinda small by itself, make no sense to put such constrains on software that meant to be portable and future proof.

Keep

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,493 @@
+}
+
+static void number_input_backspace_cb(NumberInputModel* model) {
+    uint8_t text_length = model->clear_default_text ? 1 :
```
size_t please

leedave replied:
ok

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,493 @@
+    }
+}
+
+static char* int32_to_string(int32_t value) {
```
FuriString got everything you need, why fallback to C-Strings?

leedave replied:
Ok, since I always use this to manipulate a FuriString, I'll attempt to use furi_string_printf directly in the function in place of snprintf.

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,493 @@
+    }
+}
+
+static void prevent_to_large_number(NumberInputModel* model) {
```
In terms of UI/UX this behavior is quite painful: it irreversibly modifies previous input.
Best option is to ignore input. Other systems also have visual/audial alert, but we don't use them for input.

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,493 @@
+    }
+}
+
+static void prevent_to_small_number(NumberInputModel* model) {
```
same

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,493 @@
+
+static void number_input_sign(NumberInputModel* model) {
+    if(model->sign == '-') {
+        model->sign = '+';
```
There is a model and there is a representation, what you doing is mixing representation into model.

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,493 @@
+    uint8_t text_length = model->text_buffer ? furi_string_utf8_length(model->text_buffer) : 0;
+    UNUSED(text_length);
+
+    canvas_clear(canvas);
```
clear is done by canvas before each draw call. there is no need clear in your draw call callback

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,493 @@
+    UNUSED(text_length);
+
+    canvas_clear(canvas);
+    canvas_set_color(canvas, ColorBlack);
```
same

leedave replied:
It is the same line... or are you refering to something else?

skotopes replied:
default color on draw call is black

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,493 @@
+    canvas_draw_str(canvas, 2, 9, model->header);
+    canvas_set_font(canvas, FontKeyboard);
+    // Draw keyboard
+    for(uint8_t row = 0; row < keyboard_row_count; row++) {
```
size_t please

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,493 @@
+            model->selected_row = 0;
+            model->selected_column = 0;
+            model->clear_default_text = false;
+            model->text_buffer = furi_string_alloc();
```
something really bad happens here

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,493 @@
+    with_view_model(
+        number_input->view,
+        NumberInputModel * model,
+        { furi_string_free(model->text_buffer); },
```
and here

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,493 @@
+        {
+            model->callback = callback;
+            model->callback_context = callback_context;
+            model->text_buffer = text_buffer;
```
and here

skotopes commented on applications/examples/example_number_input/example_number_input.c:
```diff
@@ -0,0 +1,77 @@
+        app->view_dispatcher,
+        ExampleNumberInputViewIdShowNumber,
+        example_number_input_show_number_get_view(app->show_number));
+    app->view_stack = view_stack_alloc();
```
I don't see you using it. Remove anything that you don't use.

skotopes commented on applications/examples/example_number_input/scenes/example_number_input_scene_input_number.c:
```diff
@@ -0,0 +1,48 @@
+    ExampleNumberInput* app = context;
+    NumberInput* number_input = app->number_input;
+    size_t enter_name_length = 5;
+    char* str = "Enter a number (0 - 9999)";
```
Numbers are clearly off

leedave replied:
The option is that you can set your own text in your app. But I will adjust this, so that the text transferred is dynamic matching the min/max numbers.

skotopes commented on applications/examples/example_number_input/scenes/example_number_input_scene_input_number.c:
```diff
@@ -0,0 +1,48 @@
+        example_number_input_scene_input_number_callback,
+        context,
+        app->text_buffer,
+        -128,
```
this

skotopes commented:
Sorry for keeping you in limbo for such a long time.

This PR was bugging me a lot when I was trying to review it previous time, It's quite bad in terms of the code and clearly requires more work before we will be able to merge it.

I've highlighted some major issues, un-draft this PR when you fix it. Also feel free to ask if you need any help.

Commit:
removed unused callback NumberChangedCallback

Commit:
change uint8_t to size_t in number_input_backspace_cb

leedave commented:
@skotopes 
Thank you for taking the time to review my code. Looks like I still have a lot to learn, but I'll stay at it. 

As chance has it, just like last time your review collided with the first day of my vacation. Meaning I can't update it before I get back. I did some of the small changes already, the rest will need to wait.

skotopes commented:
@leedave no problem, I'll wait ;-)

leedave committed:
Merge branch 'dev' into leedave/int_text_input

leedave committed:
Merge branch 'dev' into leedave/int_text_input

Commit:
Removal of unused view_stack in demo app

Commit:
copied module to app folder for faster development (remove later)

Commit:
Replaced all uint8_t with size_t... removed unused logic for selected_row < 0

Commit:
Optimize use of canvas_set_color

Commit:
Remove alloc/free of furistring that actually is a pointer

Commit:
Dynamic Header text with min/max in Example

Commit:
Removed the need of useSign in Model

Commit:
Number_input Removed sign from model, started transfer from text to int32_t

Commit:
number_input FuriString in input_show_number

Commit:
number_input FuriString in input_show_number

Commit:
limiting inputs for min/max values

Commit:
limiting inputs for min/max values

Commit:
number_input change save button on invalid numbers

Commit:
input_number update demo app to allow change of min/max

Commit:
number input fine tuning

Commit:
number_input, Remove temp development folder

Commit:
number_input, fbt format

Commit:
Bump CSV Files

leedave commented:
@skotopes 
I've done some refactoring to solve the issues you mentioned and some more. 

Main additions
- Abilitity to adjust min and max values in the demo app for easier testing of edge-cases
- Invalid number inputs are ignored
- In some cases you need to be able to write an invalid number to be able to finish off with a valid number later (min: 50, type in 9 then 0 for 90). A Lock button replaces the save button in this case to symbolize invalid input and prevent saving.
- Back button no longer saves entry, but loads back the original value
- Removed sign & useSign, as no longer needed with min/max settings
- replaced all const char* strings in module with furiString. Only used in module and destroyed on exit. 
- Callback returns number, no longer uses string

I hope this fits better. Let me know if something is still not good.

Commit:
resolve conflict in f7 csv

leedave committed:
Merge branch 'dev' into leedave/int_text_input

skotopes committed:
Merge branch 'dev' into leedave/int_text_input

skotopes commented:
![image](https://github.com/flipperdevices/flipperzero-firmware/assets/1633132/be06ca9f-b5e9-4096-9d01-dc085343fe33)
tehe

skotopes commented on applications/examples/example_number_input/scenes/example_number_input_scene_input_max.c:
```diff
@@ -0,0 +1,49 @@
+    NumberInput* number_input = app->number_input;
+
+    int32_t max = 2147483646; //has no effect if larger
+    static char str[50];
```
why static?

skotopes commented on applications/examples/example_number_input/example_number_input.c:
```diff
@@ -0,0 +1,75 @@
+    ExampleNumberInput* app = malloc(sizeof(ExampleNumberInput));
+    app->gui = furi_record_open(RECORD_GUI);
+    app->notification = furi_record_open(RECORD_NOTIFICATION);
+    app->text_buffer = furi_string_alloc();
```
dead code?

skotopes commented on applications/examples/example_number_input/example_number_input.c:
```diff
@@ -0,0 +1,75 @@
+    app->max_value = 2345;
+    furi_string_set(app->text_buffer, "5");
+    //Turn backlight on, believe me this makes testing your app easier
+    notification_message(app->notification, &sequence_display_backlight_on);
```
not needed anymore

leedave replied:
I generally hate it when loading an app via fbt and the backlight is not on. Its hard to see whats on screen without the light and it requires a keypress to activate. If the keypress already does something, you miss the initial screen. 

I'll leave it in there, but comment it out because you don't want it.

leedave replied:
Ok I just noticed this was recently changed, it now turns on without the command... so yeah, not needed anymore, will remove

skotopes commented on applications/examples/example_number_input/example_number_input.c:
```diff
@@ -0,0 +1,75 @@
+
+    scene_manager_next_scene(app->scene_manager, ExampleNumberInputSceneShowNumber);
+
+    furi_hal_power_suppress_charge_enter();
```
why?

leedave replied:
It's a habbit, as in some apps I want do things with the LED. Not needed here, so I will remove it.

skotopes commented on applications/examples/example_number_input/scenes/example_number_input_scene_input_max.c:
```diff
@@ -0,0 +1,49 @@
+    ExampleNumberInput* app = context;
+    NumberInput* number_input = app->number_input;
+
+    int32_t max = 2147483646; //has no effect if larger
```
INT32_MAX

skotopes commented on applications/examples/example_number_input/scenes/example_number_input_scene_input_min.c:
```diff
@@ -0,0 +1,49 @@
+    ExampleNumberInput* app = context;
+    NumberInput* number_input = app->number_input;
+
+    int32_t min = -2147483647; //has no effect if smaller
```
INT32_MIN

skotopes commented on applications/examples/example_number_input/scenes/example_number_input_scene_input_min.c:
```diff
@@ -0,0 +1,49 @@
+    NumberInput* number_input = app->number_input;
+
+    int32_t min = -2147483647; //has no effect if smaller
+    static char str[50];
```
why static?

skotopes commented on applications/examples/example_number_input/scenes/example_number_input_scene_input_min.c:
```diff
@@ -0,0 +1,49 @@
+
+void example_number_input_scene_input_min_on_exit(void* context) {
+    ExampleNumberInput* app = context;
+    UNUSED(app);
```
UNUSED(context)

skotopes commented on applications/examples/example_number_input/scenes/example_number_input_scene_input_number.c:
```diff
@@ -0,0 +1,48 @@
+    ExampleNumberInput* app = context;
+    NumberInput* number_input = app->number_input;
+
+    static char str[50];
```
why static?

skotopes commented on applications/examples/example_number_input/views/example_number_input_show_number.c:
```diff
@@ -0,0 +1,150 @@
+    furi_assert(context);
+}
+
+ExampleNumberInputShowNumber* example_number_input_show_number_alloc() {
```
you are inventing DialogEx here

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,487 @@
+static const size_t keyboard_origin_x = 7;
+static const size_t keyboard_origin_y = 31;
+static const size_t keyboard_row_count = 2;
+static const size_t enter_symbol = '\r';
```
those are chars

skotopes commented on applications/services/gui/modules/number_input.c:
```diff
@@ -0,0 +1,487 @@
+    number_input_draw_input(canvas, model);
+
+    canvas_set_font(canvas, FontSecondary);
+    canvas_draw_str(canvas, 2, 9, furi_string_get_cstr(model->header));
```
model->header can be null: either allocate on start or add check

skotopes commented on targets/f18/api_symbols.csv:
```diff
@@ -13,6 +13,7 @@ Header,+,applications/services/gui/icon_i.h,,
 Header,+,applications/services/gui/modules/button_menu.h,,
```
Bump API version minor

leedave replied:
Please double check if I'm doing this correctly

skotopes commented:
See comments. Un-draft when ready.

Commit:
Clear input if value is zero

Commit:
number_input: handle null on header text

Commit:
number_input: change keyboard values to char

Commit:
number input: Remove static on char for header text, change numbers to INT32_MIN/INT32_MAX

Commit:
number_input: removal of dead code

Commit:
number_input: fix for crash if number_input not opened before free

Commit:
number_input: added icon for example app

Commit:
number_input: Replaced view for show_number with DialogEx

Commit:
Number_input: FBT Format

leedave committed:
Merge branch 'dev' into leedave/int_text_input

Commit:
number_input: bump csv versions

leedave commented:
@skotopes 
Your requested updates are made, ready for review.

leedave committed:
Merge branch 'dev' into leedave/int_text_input

skotopes committed:
Merge branch 'dev' into leedave/int_text_input

skotopes commented on applications/services/gui/modules/number_input.h:
```diff
@@ -0,0 +1,68 @@
+    void* callback_context,
+    int32_t current_number,
+    int32_t min_value,
+    int32_t max_value);
```
set max value to 0 and then try to input negative number

skotopes commented:
@leedave you are close to the finishing line, just a little bit more ;-)

Commit:
number_input: allow negative input if max_value is 0

leedave commented:
Thanks again for making me a better programmer

skotopes committed:
Merge branch 'dev' into leedave/int_text_input

Commit:
Number_input: linting / format

leedave committed:
Merge branch 'dev' into leedave/int_text_input

Commit:
Removed dead code, fbt format

leedave committed:
Merge branch 'dev' into leedave/int_text_input

leedave committed:
Merge branch 'dev' into leedave/int_text_input

skotopes committed:
Merge remote-tracking branch 'origin/dev' into leedave/int_text_input

skotopes committed:
Examples: cleanup number input code

skotopes committed:
Examples: moar code cleanup in number input, simplify as much as possible, highlight incorrect input handling

skotopes commented on applications/examples/example_number_input/example_number_input.c:
```diff
@@ -0,0 +1,79 @@
+        dialog_ex_get_view(app->dialog_ex));
+
+    app->current_number = 5;
+    app->min_value = INT32_MIN;
```
I've made couple changes in example code:
- Simplified it as much as possible
- Changed min max values to int32_t specification

Looks like you incorrectly handling INT32_MIN and INT32_MAX in number_input module. Try to input something bigger or smaller that INT32_MIN or INT32_MAX.

leedave replied:
There's a reason I used +/- 1 on the INT32_MIN/INT32_MAX specifications. 

You removed these lines and replaced it with the MIN/MAX. That's why it no longer works. 
`
int32_t max = INT32_MAX - 1; // 2147483646, limit has no effect if larger
int32_t min = INT32_MIN + 1; // -2147483647 limit has no effect if smaller
`

The module needs to make an int of the input string and compare it with the min/max values. If using the max allowed int32 values the number resets if bigger/smaller and the comparison doesn't work. 

If the solution of reducing the limit by 1 is not ok, please advise me how I can fix this in a better way.

skotopes replied:
Normally you have to check `errno` for any errors that happened in conversion. However `errno` use is prohibited because it is not thread safe. Another options is to use `strtoll` and then compare int64_t values.

I've pushed bunch of fixes that solves all remaining issues.

leedave replied:
Great, thank you for the update and solution.

skotopes commented:
Un-draft when fixed

skotopes committed:
Gui: correctly handle INT_MAX and INT_MIN

skotopes committed:
Gui: fix memory leak in number input module

skotopes commented:
That was quite a journey, couple things for you to take away:
- Read documentation
- Plan object ownership and allocate/release cycle
- Keep things simple
- Test

leedave commented:
Yes that was a journey, but I do appreciate the help as at the time I started developing things for Flipper there was no documentation I could read, only other code from the FW I could learn from. 

I've just noticed that in the mean time documentation has started appearing ... and I'm surprised to see that there's actually a page related to [this Pull Request] (https://developer.flipper.net/flipperzero/doxygen/md__2github_2workspace_2applications_2examples_2example__number__input_2_read_me.html)

skotopes commented:
As long as you document your code in doxygen format it will appear in developer documentation
